icu4c/source/common/cjmlbe.cpp

```diff
@@ -0,0 +1,513 @@
+            elementList.adoptElement(new Element(ch, index), status);
+        } else {
+            // i = 1, initiate the elementList by adding six Element objects.
+            initElementList(inString, elementList, status);
+        }
+        if (U_FAILURE(status)) {
+            return 0;
+        }
+
+        evaluateBreakpoint (elementList, i, numBreaks, boundary, status);
+        if (U_FAILURE(status)) {
+            return 0;
+        }
+    }
+
+    // Add a break for the start if there is not one there already.
+    if (foundBreaks.size() == 0 || foundBreaks.peeki() < rangeStart) {
+        boundary.insertElementAt(0, 0, status);
+        numBreaks++;
+    }
+    // Add a break for the end if there is not one there already.
+    if (boundary.lastElementi() != inString.countChar32()) {
+        boundary.addElement(inString.countChar32(), status);
+        numBreaks++;
+    }
+
+    int32_t prevCPPos = -1;
+    int32_t prevUTextPos = -1;
+    int32_t correctedNumBreaks = 0;
+    for (int32_t i = 0; i < numBreaks; i++) {
+        int32_t cpPos = boundary.elementAti(i);
+        int32_t utextPos = inputMap.isValid() ? inputMap->elementAti(cpPos) : cpPos + rangeStart;
+        U_ASSERT(cpPos > prevCPPos);
+        U_ASSERT(utextPos >= prevUTextPos);
+
+        if (utextPos > prevUTextPos) {
+            if (utextPos != rangeStart
+                || (utextPos > 0
+                       && fClosePunctuationSet.contains(utext_char32At(inText, utextPos - 1)))) {
+                foundBreaks.push(utextPos, status);
+                correctedNumBreaks++;
+            }
+        }
+        else {
+            // Normalization expanded the input text, the dictionary found a boundary
+            // within the expansion, giving two boundaries with the same index in the
+            // original text. Ignore the second. See ticket #12918.
+            --numBreaks;
+        }
+        prevCPPos = cpPos;
+        prevUTextPos = utextPos;
+    }
+    (void)prevCPPos; // suppress compiler warnings about unused variable
+
+    UChar32 nextChar = utext_char32At(inText, rangeEnd);
+    if (!foundBreaks.isEmpty() && foundBreaks.peeki() == rangeEnd) {
+        // In phrase breaking, there has to be a breakpoint between Cj character and
+        // the number/open punctuation.
+        // E.g. る文字「そうだ、京都」->る▁文字▁「そうだ、▁京都」-> breakpoint between 字 and「
+        // E.g. 乗車率９０％程度だろうか -> 乗車▁率▁９０％▁程度だろうか -> breakpoint between 率 and ９
+        // E.g. しかもロゴがＵｎｉｃｏｄｅ！ -> しかも▁ロゴが▁Ｕｎｉｃｏｄｅ！-> breakpoint between が and Ｕ
+        if (!fDigitOrOpenPunctuationOrAlphabetSet.contains(nextChar)) {
+            foundBreaks.popi();
+            correctedNumBreaks--;
+        }
+    }
+
+    return correctedNumBreaks;
+}
+
+void CjMLBreakEngine::evaluateBreakpoint(const UVector &elementList, int32_t index,
+                                         int32_t &numBreaks, UVector32 &boundary,
+                                         UErrorCode &status) const {
```

Comment:
why 42, could you explain to me where that initial number come from?

Reply:
Done.
42 is previous implementation. Now, it's 26.

Reply:
This is now a named constant, MAX_FEATURE, but still without explanation.